app/models/generic_object.rb

```diff
@@ -35,31 +35,51 @@ def property_attributes
     end
   end
 
+  def property_associations
+    properties.select { |k, _| property_association_defined?(k) }.each_with_object({}) do |(k, _), h|
+      h[k] = _property_getter(k)
+    end
+  end
+
   def delete_property(name)
+    if !property_attribute_defined?(name) && !property_association_defined?(name)
+      valid_property_names = generic_object_definition.property_attributes.keys + generic_object_definition.property_associations.keys
+      raise "Invalid property [#{name}]: must be one of #{valid_property_names.join(", ")}"
+    end
+
+    val = _property_getter(name)
     properties.delete(name.to_s)
     save!
+
+    val
```

Comment:
Instead of creating a local variable you can use `tap` here to return the result of `_property_getter(name)`

```ruby
_property_getter(name).tap do
  properties.delete(name.to_s)
  save!
end
```